### ISSUE TITLE
Some type: string fields should be type: constant

### ISSUE THREAD

**Issue by eregon** [ruby/prism#1530]
Extracted from https://github.com/ruby/yarp/issues/1296 item 2 and https://github.com/ruby/yarp/issues/1134.
Note that changing from `string` to `constant` means changing in Ruby API from `String` to `Symbol` and in Java API from `byte[]` to `java.lang.String/org.jruby.RubySymbol`.
Currently there are 13 remaining `type: string` fields:

Should be `type: constant` fields (that would save a lot of serialized size + reduce C AST footprint):
https://github.com/ruby/yarp/pull/1533
- [x] CallAndWriteNode read_name, write_name
- [x] CallNode name
- [x] CallOperatorWriteNode read_name, write_name
- [x] CallOrWriteNode read_name, write_name

Less sure, I think `type: constant` would be best to save serialized size but there are other concerns:
- [ ] SymbolNode unescaped (saves serialized size if the same symbol or identifier is present multiple times in a file, likely, and can also save extra bytes -> Symbol conversions in that case).
  If we change it we should probably rename that field to `value` or so, so `unescaped` fields are always String.
  OTOH for TruffleRuby byte[] is better than java.lang.String for Symbol literals, it's a bit awkward/slow to get the bytes of a j.l.String if the symbol is not only ASCII chars or the source encoding is not UTF-8 (but feasible).
* SourceFileNode filepath (saves serialized size in case `__FILE__` is used multiple times in a file). But a bit weird that this would return a Symbol in Ruby API & for JRuby. Also would be weird to get this as a j.l.String for TruffleRuby.
  Maybe this field shouldn't exist? The caller should know what the file name is. And that has the advantage that path is only once in memory vs many times with this field (assuming multiple `__FILE__` in a file).

Should remain `type: string` fields:
* MatchLastLineNode unescaped
* RegularExpressionNode unescaped
* StringNode unescaped
* XStringNode unescaped

**Comment by eregon** [comment]
While making https://github.com/ruby/ruby/pull/8573 I noticed the last usage of string to symbol conversion (`parse_string_symbol`) is for SymbolNode. So I think we should use a a `type: constant` field for SymbolNode as detailed above.

**Comment by kddnewton** [comment]
Okay, I've done a lot of work to analyze all the possible symbols, and I think we actually need to keep it as a string, but also add an encoding flag to every string field. There's some really weird stuff, detailed in https://github.com/ruby/prism/issues/1703. For the reasons in there, I'd like to leave the symbol node as it is. In this case since the other tasks on here are closed I'm going to close this one.

**Comment by eregon** [comment]
TLDR: agreed.

I thought the encoding of a Symbol is always the same as the source, or if all 7-bit characters then it's US-ASCII.
But indeed it's not, I tried an adapted example from #1703:
```
# coding: US-ASCII
a = :"\u{C0}"
p a # => :À
p a.encoding # => #<Encoding:UTF-8>

b = :"\xef"
p b.encoding # => #<Encoding:ASCII-8BIT>
```
So indeed it's not always source encoding or US-ASCII if all-7-bit :/

I'm thinking it's very likely a symbol or identifier or the same name is present multiple times in a file so it could be a significant saving in serialized size.
e.g. `def m(kw:); end; m(kw: 42)` has a SymbolNode for the `kw: 42` part. And the RequiredKeywordParameterNode has a `:kw` constant name, so they would deduplicate.
But yeah we cannot just use `type: constant` due to that encoding concern (or we'd need to serialize the encoding for constants too, hopefully that's not needed).

Maybe we could deduplicate `string` fields too, then we'd have the encoding and avoid N duplicates of the same string in the serialization/C-structs-footprint of a given file. Anyway, that's a possible optimization for later, for correctness it needs to have the encoding.

I wonder if there is some rule for Symbol literal encoding like:
* all-7-bit -> US-ASCII
* if valid in source encoding -> source encoding
* if invalid in source encoding, UTF-8 if valid in UTF-8, BINARY otherwise
But I found a counter example:
```ruby
c = "\xC3\xA9"
p c.encoding # => #<Encoding:ASCII-8BIT>
p c.dup.force_encoding("UTF-8").valid_encoding? # true
```
So I don't think there is any rule on the unescaped value like that, it actually depends on whether `\u` is used in the literal.
